skills/repo.go

```diff
@@ -157,6 +168,211 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot grant skill to slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that skill and slot are compatible
+	if slot.Type != skill.Type {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q, skill type %q doesn't match slot type %q",
+			skillName, skillSnapName, slotName, slotSnapName, skill.Type, slot.Type)
+	}
+	// Ensure that slot and skill are not connected yet
+	if r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q twice",
+			skillName, skillSnapName, slotName, slotSnapName)
+	}
+	// Grant the skill
+	if r.skillUsedBySlot[slot] == nil {
+		r.skillUsedBySlot[slot] = make(map[*Skill]bool)
+	}
+	if r.slotsUsingSkill[skill] == nil {
+		r.slotsUsingSkill[skill] = make(map[*Slot]bool)
+	}
+	r.skillUsedBySlot[slot][skill] = true
+	r.slotsUsingSkill[skill][slot] = true
+	return nil
+}
+
+// Revoke revokes the named skill from the slot of the given snap.
+func (r *Repository) Revoke(skillSnapName, skillName, slotSnapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such skill exists
+	skill := r.skills[skillSnapName][skillName]
+	if skill == nil {
+		return fmt.Errorf("cannot revoke skill %q from snap %q, no such skill", skillName, skillSnapName)
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot revoke skill from slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that slot and skill are connected
+	if !r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot revoke skill %q from snap %q from slot %q from snap %q, it is not granted",
+			skillName, skillSnapName, slotName, slotSnapName)
+	}
+	delete(r.skillUsedBySlot[slot], skill)
+	delete(r.slotsUsingSkill[skill], slot)
+	return nil
+}
+
+// GrantedTo returns all the skills granted to a given snap.
+func (r *Repository) GrantedTo(snapName string) map[*Slot][]*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	result := make(map[*Slot][]*Skill)
+	for _, slot := range r.slots[snapName] {
+		for skill := range r.skillUsedBySlot[slot] {
+			result[slot] = append(result[slot], skill)
+		}
+		sort.Sort(bySkillSnapAndName(result[slot]))
+	}
+	return result
+}
+
+// GrantedBy returns all of the skills granted by a given snap.
+func (r *Repository) GrantedBy(snapName string) map[*Skill][]*Slot {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	result := make(map[*Skill][]*Slot)
+	for _, skill := range r.skills[snapName] {
+		for slot := range r.slotsUsingSkill[skill] {
+			result[skill] = append(result[skill], slot)
+		}
+		sort.Sort(bySlotSnapAndName(result[skill]))
+	}
+	return result
+}
+
+// UsersOf returns all of the slots using a given skill.
+func (r *Repository) UsersOf(snapName, skillName string) []*Slot {
```

Comment:
s/UsersOf/GrantsOf/

`// GrantsOf returns all of the slots that were granted the provided skill.`

This align terminology with the related functions above.